applications/MeshingApplication/custom_utilities/local_refine_tetrahedra_mesh_only_on_boundaries.hpp

```diff
@@ -0,0 +1,153 @@
+// KRATOS  __  __ _____ ____  _   _ ___ _   _  ____
+//        |  \/  | ____/ ___|| | | |_ _| \ | |/ ___|
+//        | |\/| |  _| \___ \| |_| || ||  \| | |  _
+//        | |  | | |___ ___) |  _  || || |\  | |_| |
+//        |_|  |_|_____|____/|_| |_|___|_| \_|\____| APPLICATION
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Pablo Becker
+//
+
+#pragma once
+
+// NOTE: Before compute the remeshing it is necessary to compute the neighbours
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "utilities/parallel_utilities.h"
+#include "utilities/variable_utils.h"
+
+// Applicaion includes
+#include "custom_utilities/local_refine_tetrahedra_mesh_parallel_to_boundaries.hpp"
+
+
+namespace Kratos
+{
+///@name Kratos Globals
+///@{
+///@}
+///@name Type Definitions
+///@{
+///@}
+///@name  Enum's
+///@{
+///@}
+///@name  Functions
+///@{
+///@}
+///@name Kratos Classes
+///@{
+class LocalRefineTetrahedraMeshOnlyOnBoundaries : public LocalRefineTetrahedraMeshParallelToBoundaries
+{
+public:
+
+    ///@name Type Definitions
+    ///@{
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructors
+    explicit LocalRefineTetrahedraMeshOnlyOnBoundaries(ModelPart& rModelPart)
+     : LocalRefineTetrahedraMeshParallelToBoundaries(rModelPart)
+    {
+    }
+
+    /// Destructor
+    ~LocalRefineTetrahedraMeshOnlyOnBoundaries() //TODO maybe {}
+    = default;
+
+
+    void SearchEdgeToBeRefined(
+            ModelPart& rThisModelPart,
+            compressed_matrix<int>& rCoord
+    ) override
+    {
+        KRATOS_TRY;
+        for (auto& r_elem: rThisModelPart.Elements()) {
+            if (r_elem.GetValue(SPLIT_ELEMENT)) {
+                Element::GeometryType& r_geom = r_elem.GetGeometry(); // Nodes of the element
+                for (unsigned int i = 0; i < r_geom.size(); i++) {
+                    int index_i = r_geom[i].Id() - 1;
+                    bool is_boundary_i = r_geom[i].Is(BOUNDARY);
```

Comment:
If you define these variables outside the loop it will avoid recursive allocation

Reply:
i'd rather keep it inside, makes it easier if we want to make it parallel in the future. plus the allocation size is known in compilation time, so the perfromance change wont be much